openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,344 @@
+    model : <System>
+        The system being analyzed in subproblem.
+    model_input_names : list of str or tuple
+        List of inputs requested by user to be used as inputs in the
+        subproblem's system.
+    model_output_names : list of str or tuple
+        List of outputs requested by user to be used as outputs in the
+        subproblem's system.
+    """
+
+    def __init__(self, model, inputs=None, outputs=None, driver=None, comm=None,
+                 name=None, reports=_UNDEFINED, prob_options=None, **kwargs):
+        """
+        Initialize all attributes.
+        """
+        # check for driver and issue warning about its current use
+        # in subproblem
+        if driver is not None:
+            issue_warning('Driver results may not be accurate if'
+                          ' derivatives are needed. Set driver to'
+                          ' None if your subproblem isn\'t reliant on'
+                          ' a driver.')
+
+        # make `prob_options` empty dict to be passed as **options to problem
+        # instantiation
+        if prob_options is None:
+            prob_options = {}
+
+        # call base class to set kwargs
+        super().__init__(**kwargs)
+
+        # store inputs and outputs in options
+        self.options.declare('inputs', {}, types=dict,
+                             desc='Subproblem Component inputs')
+        self.options.declare('outputs', {}, types=dict,
+                             desc='Subproblem Component outputs')
+
+        # set other variables necessary for subproblem
+
+        self.prob_args = {'driver': driver,
+                          'comm': comm,
+                          'name': name,
+                          'reports': reports}
+
+        self.prob_args.update(prob_options)
+
+        self.model = model
+        self.model_input_names = inputs
+        self.model_output_names = outputs
+
+    def setup(self):
+        """
+        Perform some final setup and checks.
+        """
+        p = self._subprob = Problem(**self.prob_args)
+        p.model.add_subsystem('subsys', self.model, promotes=['*'])
+
+        # perform first setup to be able to get inputs and outputs
+        p.setup(force_alloc_complex=self._problem_meta['force_alloc_complex'])
+        p.final_setup()
+
+        # boundary inputs are any inputs that externally come into `SubproblemComp`
+        boundary_inputs = p.model.list_inputs(out_stream=None, prom_name=True,
+                                              units=True, shape=True, desc=True, is_indep_var=True)
+        # want all outputs from the `SubproblemComp`
+        all_outputs = p.model.list_outputs(out_stream=None, prom_name=True,
+                                           units=True, shape=True, desc=True)
+
+        # declaring all inputs as design vars and all outputs as constraints allows for coloring
+        # to be computed
+        for _, meta in boundary_inputs:
```

Comment:
Isn't setting all inputs as DVs and all outputs as constraints going to mess us up later when the driver is an actual optimizer, i.e. it might only be using some subset of the variables as dv/obj/constraints, so it seems like we need to keep track of what dv/obj/constraints exist as part of the model and be able to save/restore them after we have the coloring, or (maybe better) we update the coloring stuff to work with a caller supplied list of 'of' and 'wrt' vars.

Reply:
Made methods for `add_design_vars`, `add_constraints`, and `add_objectives` and made `_setup_var_data` to do coloring after we get design vars, constraints, and objectives